tests/src/Interop/IJW/NativeVarargs/IjwNativeVarargs.cpp

```diff
@@ -0,0 +1,474 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+#include <cstddef>
+#include <vector>
+#include <cstdarg>
+#include <numeric>
+#include <array>
+#include <functional>
+#include <iostream>
+#using <mscorlib.dll>
+using namespace System::Collections::Generic;
+
+public enum class TestCases
+{
+    SumInts,
+    SumFloats,
+    SumLongLongs,
+    SumPromotedShorts,
+    SumDoubles,
+    SumPromotedFloats,
+    SumHFAs
+};
+
+struct HFA
+{
+    float f1;
+    float f2;
+    float f3;
+    float f4;
+};
+
+#pragma unmanaged
+
+int SumInts(std::size_t numElements, ...)
+{
+    va_list args;
+    va_start(args, numElements);
+    int sum = 0;
+    for (std::size_t i = 0; i < numElements; ++i)
+    {
+        sum += va_arg(args, int);
+    }
+    va_end(args);
+    return sum;
+}
+
+long long SumLongLongs(std::size_t numElements, ...)
+{
+    va_list args;
+    va_start(args, numElements);
+    long long sum = 0;
+    for (std::size_t i = 0; i < numElements; ++i)
+    {
+        sum += va_arg(args, long long);
+    }
+    va_end(args);
+    return sum;
+}
+
+double SumDoubles(std::size_t numElements, ...)
+{
+    va_list args;
+    va_start(args, numElements);
+    double sum = 0;
+    for (std::size_t i = 0; i < numElements; ++i)
+    {
+        sum += va_arg(args, double);
+    }
+    va_end(args);
+    return sum;
+}
+
+float SumHFAs(std::size_t numElements, ...)
+{
+    va_list args;
+    va_start(args, numElements);
+    float sum = 0;
+    for (std::size_t i = 0; i < numElements; ++i)
+    {
+        HFA hfa = va_arg(args, HFA);
+        sum += hfa.f1 + hfa.f2 + hfa.f3 + hfa.f4;
+    }
+    va_end(args);
+    return sum;
+}
+
+constexpr std::size_t NumArgsPerCall = 41; // Match number of arguments used in JIT/Directed/arglist/vararg test.
+
+#pragma managed
+
+public ref class TestClass
+{
+public:
+    List<TestCases>^ RunTests(int seed)
+    {
+        System::Random^ rng = gcnew System::Random(seed);
```

Comment:
Normally random number generators are frowned on in tests as they do not produce reproducible data. I think I have the same preference to avoid them.

Reply:
We pass in a consistent seed from the managed side of the test.